### PR TITLE
ref: senderprovider for each client

### PR DESCRIPTION
**Pull request by lasrinnil**
## Description

In our earlier implementation, we had a ServiceBusRequestSender which was registered with a reference to the queue in which the reply was expected. This worked when we only used the DefaultChargeLinkClient, but now we use both clients (DefaultChargeLinkClient, DefaultChargeLinkMessagesClient) we have no way to register a way for both clients to have their own reply.

This PR will introduce a provider for each client to resolve a sender with their own reply.

## References

* #784

### CODE REVIEW

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.IntegrationTests/DefaultChargeLink/DefaultChargeLinkTests.cs [hunk 1]
```diff
@@ -35,25 +35,25 @@ public static class DefaultChargeLinkTests
...
             public CreateDefaultChargeLinksRequestAsync(ChargesClientsFixture fixture, ITestOutputHelper testOutputHelper)
                 : base(fixture, testOutputHelper)
             {
                 _replyToQueueName = EnvironmentVariableReader.GetEnvironmentVariable(
                     EnvironmentSettingNames.CreateLinkReplyQueueName, string.Empty);
-                _requestQueueName = EnvironmentVariableReader.GetEnvironmentVariable(
+                var requestQueueName = EnvironmentVariableReader.GetEnvironmentVariable(
                     EnvironmentSettingNames.CreateLinkRequestQueueName, string.Empty);
 
                 string serviceBusConnectionString = EnvironmentVariableReader.GetEnvironmentVariable(
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.IntegrationTests/DefaultChargeLink/DefaultChargeLinkTests.cs:50]: We should not use env variables in code except in configuration.

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkRequestClientTests.cs [hunk 2]
```diff
@@ -61,29 +58,15 @@ public class DefaultChargeLinkClientTests
         [Theory]
         [InlineAutoDomainData]
         public async Task CreateDefaultChargeLinksRequestAsync_WhenInputIsValid_SendsMessage(
-            [NotNull] [Frozen] Mock<ServiceBusClient> serviceBusClientMock,
-            [NotNull] [Frozen] Mock<IServiceBusRequestSenderFactory> serviceBusRequestSenderFactoryMock,
-            [NotNull] [Frozen] Mock<IServiceBusRequestSender> serviceBusRequestSenderMock,
-            string createLinkRequestQueueName)
+            [NotNull] [Frozen] Mock<IServiceBusRequestSenderProvider> serviceBusRequestSenderFactoryMock,
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkRequestClientTests.cs:61]:
```suggestion
            [NotNull] [Frozen] Mock<IServiceBusRequestSenderProvider> serviceBusRequestSenderProviderMock,
```

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkRequestClientTests.cs [hunk 3]
```diff
@@ -95,145 +78,144 @@ public class DefaultChargeLinkClientTests
...
-
-            // Assert
-            serviceBusRequestSenderMock.Verify(
-                x => x.SendRequestAsync(
-                    It.IsAny<byte[]>(),
-                    replyQueueName,
-                    CorrelationId),
-                Times.Once);
-        }
+        // [Theory]
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkRequestClientTests.cs:85]: Remove uncommented code?

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Energinet.DataHub.Charges.Clients.csproj [hunk 4]
```diff
@@ -29,7 +29,7 @@ limitations under the License.
   <!-- Configuration for NuGet package -->
   <PropertyGroup>
     <PackageId>Energinet.DataHub.Charges.Clients</PackageId>
-    <PackageVersion>1.0.6$(VersionSuffix)</PackageVersion>
+    <PackageVersion>1.0.7$(VersionSuffix)</PackageVersion>
```
Comment (HenrikSommer) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Energinet.DataHub.Charges.Clients.csproj:32]: Remember to add release note in release-notes.md

**Review by HenrikSommer — APPROVED** [review]

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLinkMessages/DefaultChargeLinkMessagesClientTests.cs [hunk 5]
```diff
@@ -45,13 +44,12 @@ public class DefaultChargeLinkMessagesClientTests
             string meteringPointId,
             string correlationId,
             [NotNull] [Frozen] Mock<ServiceBusClient> serviceBusClientMock,
-            [NotNull] [Frozen] Mock<IServiceBusRequestSenderFactory> serviceBusRequestSenderFactoryMock)
+            [NotNull] [Frozen] Mock<IServiceBusRequestSenderProvider> serviceBusRequestSenderFactoryMock)
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLinkMessages/DefaultChargeLinkMessagesClientTests.cs:47]:
```suggestion
            [NotNull] [Frozen] Mock<IServiceBusRequestSenderProvider> serviceBusRequestSenderProviderMock)
```
Will not report more of this naming issue.

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLinkMessages/DefaultChargeLinkMessagesClientTests.cs [hunk 6]
```diff
@@ -98,146 +85,145 @@ public class DefaultChargeLinkMessagesClientTests
...
-
-            // Assert
-            serviceBusRequestSenderMock.Verify(
-                x => x.SendRequestAsync(
-                    It.IsAny<byte[]>(),
-                    anyReplyQueue,
-                    anyCorrelationId),
-                Times.Once);
-        }
+        // [Theory]
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLinkMessages/DefaultChargeLinkMessagesClientTests.cs:92]: Remove?

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkRequestClientTests.cs (hunk 3 above)
Comment (lasrinnil) [reply source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkRequestClientTests.cs:85]: reintroduced them

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/ServiceBus/ServiceBusRequestSenderTests.cs [hunk 7]
```diff
@@ -46,12 +46,12 @@ public class ServiceBusRequestSenderTests
...
             serviceBusSenderMock.Setup(x => x.SendMessageAsync(It.IsAny<ServiceBusMessage>(), default));
 
-            var sut = new ServiceBusRequestSender(serviceBusClientMock.Object, replyToQueueName);
+            var sut = new ServiceBusRequestSender(serviceBusSenderMock.Object, replyToQueueName);
 
             var defaultChargeLinks = new CreateDefaultChargeLinks { MeteringPointId = meteringPointId };
 
             // Act
-            await sut.SendRequestAsync(defaultChargeLinks.ToByteArray(), requestQueueName, correlationId).ConfigureAwait(false);
+            await sut.SendRequestAsync(defaultChargeLinks.ToByteArray(), correlationId).ConfigureAwait(false);
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients.Tests/ServiceBus/ServiceBusRequestSenderTests.cs:54]: I'm removing correlation ID all over in the production code.
Could you perhaps do the same? Or should we put it on dev session board?

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Providers/DefaultChargeLinkMessagesClientServiceBusRequestSenderProvider.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+using Azure.Messaging.ServiceBus;
+
+namespace Energinet.DataHub.Charges.Libraries.Providers
+{
+    public class DefaultChargeLinkMessagesClientServiceBusRequestSenderProvider : ServiceBusRequestSenderProviderBase, IServiceBusRequestSenderProvider
+    {
+        public DefaultChargeLinkMessagesClientServiceBusRequestSenderProvider(
+            ServiceBusClient serviceBusClient,
+            string replyToQueueName,
+            string requestQueueName = "create-link-messages-request")
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Providers/DefaultChargeLinkMessagesClientServiceBusRequestSenderProvider.cs:24]: Is this the only location of the queue name? Would it make sense to make a collection of const strings of infrastructure names?

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/ServiceBus/ServiceBusRequestSender.cs [hunk 9]
```diff
@@ -22,20 +22,18 @@ namespace Energinet.DataHub.Charges.Libraries.ServiceBus
...
-        public async Task SendRequestAsync([NotNull] byte[] data, [NotNull] string requestQueueName, [NotNull] string correlationId)
+        public async Task SendRequestAsync([NotNull] byte[] data, [NotNull] string correlationId)
         {
-            await using var sender = _serviceBusClient.CreateSender(requestQueueName);
-
-            await sender.SendMessageAsync(new ServiceBusMessage
+            await _serviceBusSender.SendMessageAsync(new ServiceBusMessage
             {
                 Body = new BinaryData(data),
                 ApplicationProperties = { new KeyValuePair<string, object>("ReplyTo", _replyToQueueName) },
```
Comment (BjarkeMeier) [inline source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/ServiceBus/ServiceBusRequestSender.cs:39]: "ReplyTo" is defined at the system level. Would it be helpful to put it somewhere as a const string?

**Review by BjarkeMeier — APPROVED** [review]
Nice job.
Package documentation on how to configure the container would be very useful.

File: source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Providers/DefaultChargeLinkMessagesClientServiceBusRequestSenderProvider.cs (hunk 8 above)
Comment (lasrinnil) [reply source/Energinet.Charges.Libraries/source/Energinet.DataHub.Charges.Clients/Providers/DefaultChargeLinkMessagesClientServiceBusRequestSenderProvider.cs:25]: In upcoming work, we are focusing on move parts that are shared closer together, I'm making a note of this and will implement it.

**Pull request merged**